scvi/models/_base.py

```diff
@@ -0,0 +1,792 @@
+        for tensors in post:
+            x = tensors[_CONSTANTS.X_KEY]
+            batch_idx = tensors[_CONSTANTS.BATCH_KEY]
+            labels = tensors[_CONSTANTS.LABELS_KEY]
+            outputs = self.model.inference(
+                x, batch_index=batch_idx, y=labels, n_samples=n_samples
+            )
+            px_r = outputs["px_r"]
+            px_rate = outputs["px_rate"]
+            px_dropout = outputs["px_dropout"]
+
+            if self.model.gene_likelihood == "poisson":
+                l_train = px_rate
+                l_train = torch.clamp(l_train, max=1e8)
+                dist = torch.distributions.Poisson(
+                    l_train
+                )  # Shape : (n_samples, n_cells_batch, n_genes)
+            elif self.model.gene_likelihood == "nb":
+                dist = NegativeBinomial(mu=px_rate, theta=px_r)
+            elif self.model.gene_likelihood == "zinb":
+                dist = ZeroInflatedNegativeBinomial(
+                    mu=px_rate, theta=px_r, zi_logits=px_dropout
+                )
+            else:
+                raise ValueError(
+                    "{} reconstruction error not handled right now".format(
+                        self.model.gene_likelihood
+                    )
+                )
+            if n_samples > 1:
+                exprs = dist.sample().permute(
+                    [1, 2, 0]
+                )  # Shape : (n_cells_batch, n_genes, n_samples)
+            else:
+                exprs = dist.sample()
+
+            if gene_list is not None:
+                exprs = exprs[:, gene_mask, ...]
+
+            x_new.append(exprs.cpu())
+        x_new = torch.cat(x_new)  # Shape (n_cells, n_genes, n_samples)
+
+        return x_new.numpy()
+
+    @torch.no_grad()
+    def _get_denoised_samples(
+        self,
+        adata=None,
+        indices=None,
+        n_samples: int = 25,
+        batch_size: int = 64,
+        rna_size_factor: int = 1000,
+        transform_batch: Optional[int] = None,
+    ) -> np.ndarray:
+        """
+        Return samples from an adjusted posterior predictive.
+
+        Parameters
+        ----------
+        adata
+            AnnData object that has been registered with scvi. If `None`, defaults to the
+            AnnData object used to initialize the model.
+        indices
+            Indices of cells in adata to use. If `None`, all cells are used.
+        n_samples
+            How may samples per cell
+        batch_size
+            Minibatch size for data loading into model
+        rna_size_factor
+            size factor for RNA prior to sampling gamma distribution
+        transform_batch
+            int of which batch to condition on for all cells
+
+        Returns
+        -------
+        denoised_samples
+        """
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
+
+        posterior_list = []
+        for tensors in post:
```

Comment:
this looks like code duplicate with the posterior predictive density. Can we refactor that?

Reply:
yea, will include in another PR